File: modules/traditional-rag/src/traditional_rag/http/routes/sources.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel, Field

from traditional_rag.core.sources import create_source, ensure_default_private_source, get_readable_source, list_readable_sources, update_source
from traditional_rag.core.types import UserContext
from traditional_rag.http.auth import require_internal_user
from traditional_rag.http.errors import raise_http_error
from traditional_rag.http.serializers import to_public_source

router = APIRouter(prefix="/traditional")


class CreateSourceRequest(BaseModel):
    name: str = Field(min_length=3, max_length=96)
    kind: str = Field(default="private", pattern="^(private|public)$")
    description: str = Field(default="", max_length=1000)


class UpdateSourceRequest(BaseModel):
    name: str | None = Field(default=None, min_length=3, max_length=96)
    description: str | None = Field(default=None, max_length=1000)


@router.get("/sources")
async def list_sources_route(user: UserContext = Depends(require_internal_user)) -> dict:
    try:
        if not user.is_admin:
            await ensure_default_private_source(user.user_id, user.username)
        sources = await list_readable_sources(user)
        return {"sources": [to_public_source(source) for source in sources]}
    except Exception as error:
        raise_http_error(error)


@router.post("/sources", status_code=201)
async def create_source_route(body: CreateSourceRequest, user: UserContext = Depends(require_internal_user)) -> dict:
    try:
        source = await create_source(user, body.name, body.kind, body.description)
        return {"source": to_public_source(source)}
    except Exception as error:
        raise_http_error(error)


@router.get("/sources/{source_id}")
async def get_source_route(source_id: str, user: UserContext = Depends(require_internal_user)) -> dict:
    try:
        source = await get_readable_source(user, source_id)
        return {"source": to_public_source(source)}
    except Exception as error:
        raise_http_error(error)


@router.patch("/sources/{source_id}")
async def update_source_route(source_id: str, body: UpdateSourceRequest, user: UserContext = Depends(require_internal_user)) -> dict:
    try:
        source = await update_source(user, source_id, name=body.name, description=body.description)
        return {"source": to_public_source(source)}
    except Exception as error:
        raise_http_error(error)

File: modules/nano-brain/src/http/routes/sources.ts
import { Hono } from 'hono';
import { createPublicSource, getReadableSource, listReadableSources, updateSourceName } from '../../core/sources';
import { internalRoute } from '../route';
import { toPublicSource } from '../serializers';

export const sourcesRouter = new Hono();

sourcesRouter.get(
  '/nano/sources',
  internalRoute('读取 source 列表失败', async (c, ctx) => {
    const sources = await listReadableSources(ctx);
    return c.json({ sources: sources.map(toPublicSource) });
  }),
);

sourcesRouter.post(
  '/nano/sources',
  internalRoute('创建 source 失败', async (c, ctx) => {
    const body = await c.req.json();
    const source = await createPublicSource(ctx, { name: body.name });
    return c.json({ source: toPublicSource(source) }, 201);
  }),
);

sourcesRouter.get(
  '/nano/sources/:sourceId',
  internalRoute('读取 source 失败', async (c, ctx) => {
    const source = await getReadableSource(ctx, c.req.param('sourceId')!);
    return c.json({ source: toPublicSource(source) });
  }),
);

sourcesRouter.patch(
  '/nano/sources/:sourceId',
  internalRoute('更新 source 失败', async (c, ctx) => {
    const body = await c.req.json();
    const source = await updateSourceName(ctx, c.req.param('sourceId')!, { name: body.name });
    return c.json({ source: toPublicSource(source) });
  }),
);
